Source/Core/VideoCommon/PerformanceTracker.cpp

```diff
@@ -0,0 +1,252 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "VideoCommon/PerformanceTracker.h"
+
+#include <algorithm>
+#include <cmath>
+#include <implot.h>
+#include <iomanip>
+
+#include "Common/CommonTypes.h"
+#include "Common/FileUtil.h"
+#include "Common/Timer.h"
+#include "Core/Core.h"
+#include "VideoCommon/VideoConfig.h"
+
+static constexpr double SAMPLE_RC_RATIO = 0.25;
+
+PerformanceTracker::PerformanceTracker(const char* log_name,
+                                       const std::optional<s64> sample_window_us)
+    : m_on_state_changed_handle{Core::AddOnStateChangedCallback([this](Core::State state) {
+        if (state == Core::State::Paused)
+          SetPaused(true);
+        else if (state == Core::State::Running)
+          SetPaused(false);
+      })},
+      m_log_name{log_name}, m_sample_window_us{sample_window_us}
+{
+  Reset();
+}
+
+PerformanceTracker::~PerformanceTracker()
+{
+  Core::RemoveOnStateChangedCallback(&m_on_state_changed_handle);
+}
+
+void PerformanceTracker::Reset()
+{
+  std::lock_guard lock{m_mutex};
+
+  QueueClear();
+  m_last_time = Clock::now();
+  m_hz_avg = 0.0;
+  m_dt_avg = DT::zero();
+  m_dt_std = std::nullopt;
+}
+
+void PerformanceTracker::Count()
+{
+  std::lock_guard lock{m_mutex};
+
+  if (m_paused)
+    return;
+
+  const DT window{GetSampleWindow()};
+
+  const TimePoint time{Clock::now()};
+  const DT diff{time - m_last_time};
+
+  m_last_time = time;
+
+  QueuePush(diff);
+  m_dt_total += diff;
+
+  if (m_dt_queue_begin == m_dt_queue_end)
+    m_dt_total -= QueuePop();
+
+  while (window <= m_dt_total - QueueTop())
+    m_dt_total -= QueuePop();
+
+  // Simple Moving Average Throughout the Window
+  m_dt_avg = m_dt_total / QueueSize();
+  const double hz = DT_s(1.0) / m_dt_avg;
+
+  // Exponential Moving Average
+  const DT_s rc = SAMPLE_RC_RATIO * std::min(window, m_dt_total);
+  const double a = 1.0 - std::exp(-(DT_s(diff) / rc));
+
+  // Sometimes euler averages can break when the average is inf/nan
+  if (std::isfinite(m_hz_avg))
+    m_hz_avg += a * (hz - m_hz_avg);
+  else
+    m_hz_avg = hz;
+
+  m_dt_std = std::nullopt;
+
+  if (m_log_name && g_ActiveConfig.bLogRenderTimeToFile)
+    LogRenderTimeToFile(diff);
+}
+
+DT PerformanceTracker::GetSampleWindow() const
+{
+  // This reads a constant value and thus does not need a mutex
+  return std::chrono::duration_cast<DT>(
+      DT_us(m_sample_window_us.value_or(std::max(1, g_ActiveConfig.iPerfSampleUSec))));
+}
+
+double PerformanceTracker::GetHzAvg() const
+{
+  std::lock_guard lock{m_mutex};
```

Comment:
A bunch of these could be reader locks (std::shared_lock) to save a bit on contention. In practice it's unlikely to make a difference here, this seems like it's mostly single reader / single writer in the current workload. But using reader locks doesn't make anything more complicated, so in general I prefer doing it regardless.

std::shared_mutex is also marginally better in general for various historical reasons: https://stackoverflow.com/questions/69990339/why-is-stdmutex-so-much-worse-than-stdshared-mutex-in-visual-c